src/runtime.cc

```diff
@@ -224,17 +232,27 @@ namespace detail {
 		m_d_queue->wait();
 		m_h_queue->wait();
 
-		if(spdlog::should_log(log_level::trace) && m_cfg->is_recording()) {
-			if(m_local_nid == 0) { // It's the same across all nodes
-				assert(m_task_recorder.get() != nullptr);
-				const auto graph_str = detail::print_task_graph(*m_task_recorder);
-				CELERITY_TRACE("Task graph:\n\n{}\n", graph_str);
+		if(m_cfg->is_recording()) {
+			if(spdlog::should_log(log_level::trace)) {
+				if(m_local_nid == 0) { // It's the same across all nodes
+					assert(m_task_recorder.get() != nullptr);
+					const auto graph_str = detail::print_task_graph(*m_task_recorder);
+					CELERITY_TRACE("Task graph:\n\n{}\n", graph_str);
+				}
+				// must be called on all nodes
+				auto cmd_graph = gather_command_graph();
+				if(m_local_nid == 0) {
+					std::this_thread::sleep_for(std::chrono::milliseconds(500)); // Avoid racing on stdout with other nodes (funneled through mpirun)
+					CELERITY_TRACE("Command graph:\n\n{}\n", cmd_graph);
+				}
 			}
-			// must be called on all nodes
-			auto cmd_graph = gather_command_graph();
-			if(m_local_nid == 0) {
-				std::this_thread::sleep_for(std::chrono::milliseconds(500)); // Avoid racing on stdout with other nodes (funneled through mpirun)
-				CELERITY_TRACE("Command graph:\n\n{}\n", cmd_graph);
+
+			if(m_divergence_check != nullptr) {
+				// Sychronize all nodes before reseting shuch that we don't get into a deadlock
```

Comment:
```suggestion
				// Sychronize all nodes before resetting such that we don't get into a deadlock
```
Could you briefly explain why we could deadlock here?